{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "output_file = \"overlap_results_SNP.tsv\"\n",
    "fout = open(output_file, 'w')\n",
    "\n",
    "class Interval:\n",
    "\n",
    "    def __init__(self, chrom, start, end, feature, gene_id=None, gene_name=None):\n",
    "        '''\n",
    "        Constructor, containing features representing an area on the chromosome.\n",
    "        chrom: chromosome\n",
    "        start: Starting position of the area on chr.\n",
    "        end: Ending position of the area on chr.\n",
    "        feature: Type of genetic feature represented by this area (e.g., \"gene\", \"exon\").\n",
    "        gene_id: Gene identifier.\n",
    "        gene_name: Gene name.\n",
    "        '''\n",
    "        self.chrom = chrom\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.feature = feature\n",
    "        self.gene_id = gene_id\n",
    "        self.gene_name = gene_name\n",
    "\n",
    "    def overlaps(self, other):\n",
    "        # Does this area on the chromosome overlap with another area on the chromosome?\n",
    "        return self.chrom == other.chrom and self.end >= other.start and self.start <= other.end\n",
    "\n",
    "    def distance_to(self, other):\n",
    "        # Calculate the distance between this area on the chromosome and another area on the chromosome.\n",
    "        # If areas on the chromosome overlap, the distance between them is 0.\n",
    "        if self.chrom != other.chrom:\n",
    "            return float('inf')\n",
    "        if self.end < other.start:\n",
    "            return other.start - self.end\n",
    "        elif other.end < self.start:\n",
    "            return self.start - other.end\n",
    "        else:\n",
    "            return 0\n",
    "\n",
    "def extract_gene_info(attributes):\n",
    "    # Extract gene identifier and name information from attributes.\n",
    "    gene_id_match = re.search('gene_id \"([^\"]+)\"', attributes)\n",
    "    gene_name_match = re.search('gene_name \"([^\"]+)\"', attributes)\n",
    "\n",
    "    gene_id = gene_id_match.group(1) if gene_id_match else None\n",
    "    gene_name = gene_name_match.group(1) if gene_name_match else None\n",
    "\n",
    "    return gene_id, gene_name\n",
    "\n",
    "# Load the GTF file with the GRCh38 version of the genome\n",
    "gtf_file = \"Homo_sapiens.GRCh38.110.gtf\"\n",
    "\n",
    "# Dictionary of genetic features\n",
    "features_dict = {\n",
    "    \"exon\": [],\n",
    "    \"transcript\": [],\n",
    "    \"CDS\": [],\n",
    "    \"start_codon\": [],\n",
    "    \"stop_codon\": [],\n",
    "    \"three_prime_utr\": [],\n",
    "    \"five_prime_utr\": [],\n",
    "    \"gene\": []    \n",
    "}\n",
    "coding_genes = set()\n",
    "\n",
    "\n",
    "with open(gtf_file, 'r') as fin:\n",
    "    # Open GTF file\n",
    "    for line in fin:\n",
    "        if not line.startswith('#'):  # Skip comments\n",
    "            fields = line.strip().split('\\t')\n",
    "            if len(fields) != 9:\n",
    "                print(f\"Unexpected line format: {line}\")\n",
    "                continue\n",
    "\n",
    "            # Assign attributes\n",
    "            chrom = fields[0]\n",
    "            start = int(fields[3]) - 1\n",
    "            end = int(fields[4])\n",
    "            feature = fields[2]\n",
    "            gene_id, gene_name = extract_gene_info(fields[8])\n",
    "\n",
    "            if feature == \"CDS\":\n",
    "                # Is the processed genetic feature \"CDS\"?\n",
    "                coding_genes.add(gene_id)\n",
    "\n",
    "            if feature in features_dict:\n",
    "                # Does the processed genetic feature correspond to one in the dictionary?\n",
    "                features_dict[feature].append(Interval(chrom, start, end, feature, gene_id, gene_name))\n",
    "\n",
    "input_file = \"output_results_SNP.tsv\"\n",
    "\n",
    "with open(input_file, 'r') as fin:\n",
    "    # Open file with rsIDs and coordinates\n",
    "    headers = next(fin).strip().split('\\t')\n",
    "    rsid_index = headers.index('rsID')\n",
    "    location_index = headers.index('Location')\n",
    "\n",
    "    # For each rsID, find genetic features it overlaps with\n",
    "    for line in fin:\n",
    "        fields = line.strip().split('\\t')\n",
    "        # get rsID and coordinates from input file\n",
    "        rsid = fields[rsid_index]\n",
    "        location = fields[location_index]\n",
    "        chrom, pos = location.split(':')\n",
    "        start, end = map(int, pos.split('-'))\n",
    "\n",
    "        # Create Interval object for rsID coordinates\n",
    "        coord_interval = Interval(chrom, start, end, \"Query\")\n",
    "        \n",
    "        overlapping_features = set()\n",
    "        \n",
    "        # For each genetic feature in the dictionary of genetic features, find features overlapping with rsID\n",
    "        for feature, intervals in features_dict.items():\n",
    "        #Iterate through the dictionary of genetic features\n",
    "            for interval in intervals:\n",
    "                if interval.overlaps(coord_interval):\n",
    "                #If features overlap,, add them to the set of features overlapping with rsID\n",
    "                    if interval.gene_name:\n",
    "                    #If gene name is known, add it to the set of features overlapping with rsID\n",
    "                        overlapping_features.add(f\"{feature}_{interval.gene_name}\")\n",
    "                    elif interval.gene_id:\n",
    "                    #If gene name is not known, but its identifier is known, add it to the set of features overlapping with rsID\n",
    "                        overlapping_features.add(f\"{feature}_{interval.gene_id}\")\n",
    "                    else:\n",
    "                        overlapping_features.add(feature)\n",
    "                        \n",
    "        nearest_gene_distance = float('inf')\n",
    "        nearest_gene_info = None\n",
    "        for gene in features_dict[\"gene\"]:\n",
    "            #Iterate through the list of coding genes\n",
    "            if gene.gene_id in coding_genes:\n",
    "                #Is the gene coding?\n",
    "                distance = coord_interval.distance_to(gene)\n",
    "                #Calculate the distance between rsID and the gene\n",
    "                if distance < nearest_gene_distance:\n",
    "                    #If the distance is smaller than the smallest distance, update the information about the nearest gene\n",
    "                    nearest_gene_distance = distance\n",
    "                    #If the gene name is known, add it to the set of features overlapping with rsID\n",
    "                    nearest_gene_info = gene.gene_name if gene.gene_name else gene.gene_id\n",
    "\n",
    "        # Write to file rsID, features overlapping with rsID, nearest coding gene\n",
    "        overlap_str = ', '.join(overlapping_features) if overlapping_features else \"No direct overlap\"\n",
    "        nearest_gene_str = nearest_gene_info if nearest_gene_info else \"No nearby coding gene found\"\n",
    "        fout.write(f\"{rsid} overlaps with: {overlap_str}. Nearest coding gene: {nearest_gene_str}\\n\")\n",
    "\n",
    "fout.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
